### PR TITLE
Add dev guides for modifiers and package managers

### PR DESCRIPTION
**Pull request by douglasjacobsen**

### CODE REVIEW

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+Modifier definition files are stored within object repositories. These
+repositories generally store modifiers within a directory named ``modifiers``.
+Individual repositories can control this through their own config file (
+``repo.yaml`` or ``modifier_repo.yaml`` ).
+
+Within the repository, each modifier definition file is a python module that is
+stored within a directory named for the modifier. As an example, Ramble comes
+with a repository named ``builtin``. This repository contains several standard
+modifier definitions that are proided to the community. One of the modifier
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]: provided

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+#. Commands you need to execute as part of your changes
+#. Environment variables that need to be changed (set, appended, prepended, or unset)
+#. An example of output the modifier would create (if any)
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Compilation / Installation
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some modifiers require specific software packages to be available to function
+properly. Modifiers can define software packages they both will install with
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]: "they both install..." somehow sounds a bit weird ;)

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+``register_phase`` directive can be used to add a phase into a pipeline. Phases
+are written as class methods with a specific signature, and will be
+automatically executed as part of the pipeline they belong to.
+
+^^^^^^^^^^^^^^^^^^^^^^^
+Executable Modification
+^^^^^^^^^^^^^^^^^^^^^^^
+
+One of the most powerful modifications available within modifier definitions is
+the ``executable_modification`` directive. Some modifiers will require the
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]: Is it `executable_modifiers`?

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+ability to inject commands around commands that exist in the experiments
+already. A good example of this is a performance analysis tool, which needs to
+modify the execution command to profile the experiment, and generate a summary
+of the performance characteristics of the experiment after it is complete.
+
+To accomplish this goal, the ``executable_modification`` directive can be used,
+which is implemented as a class method which returns two lists of
+``CommandExecutable`` objects, which are injected before and after the each
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]: Redundant "the"

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+Some modifiers need to edit environment variables within an experiment. The
+``env_var_modification`` directive can be used to change existing environment
+variables.
+
+^^^^^^^^^^^^^^^^^^^^
+Shared Functionality
+^^^^^^^^^^^^^^^^^^^^
+
+This are several directives that are shared between applications and modifiers.
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]: "There are"

File: lib/ramble/docs/dev_guides/package_manager_dev_guide.rst [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+Existing package manager classes can be referenced using the:
+``from ramble.pkg_man.builtin.<package_manager_name> import <package_manager_class>``
+syntax.
+
+------------------------------------
+Writing a package manager definition
+------------------------------------
+
+After a package managers's ``package_manager.py`` file is created, Ramble's
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/package_manager_dev_guide.rst]: Extra "s"

File: lib/ramble/docs/dev_guides/package_manager_dev_guide.rst [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+useful for a package manager to define a variable which users can modify to
+control the package manager's behavior. The ``package_manager_variable``
+directive can be used to define these variables.
+
+^^^^^^^^
+Builtins
+^^^^^^^^
+
+A builtin is a specific command that is injected within an experiment. Builtins
+can have dependencies, and and be injected either at the beginning or end of
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/package_manager_dev_guide.rst]: "and" -> "can"

File: lib/ramble/docs/dev_guides/package_manager_dev_guide.rst [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+:ref:`variants <variants-config>` configuration section. Specifically
+``variants:package_manager:<package_manager_name>``. This quantity can be
+parameterized, and can refer to a variable which defines which package manager
+is used.
+
+The behavior of a package manager when executing a ``dry-run`` depends on how
+the runner and package manager were implemented. It is recommended that each
+individual runner implement its own ``dry-run`` behavior to enable exploration
+with the package manager without having to perform full installation. Existing
+package managers can be used as a reference.
```
Comment (linsword13) [inline lib/ramble/docs/dev_guides/package_manager_dev_guide.rst:172]: I don't know if it'd be worthwhile to mention the existing unit-test support in `known_applications` for dry-run test package managers?

**Review by linsword13 — COMMENTED** [review]
Thanks for writing these up, super informative! Some nits and typos but otherwise LGTM.

**Comment by douglasjacobsen** [comment]
Thanks for catching those. Fixed in the latest commit.

**Review by linsword13 — DISMISSED** [review]

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+   option. This file may not be copied, modified, or distributed
+   except according to those terms.
+
+.. _modifier-dev-guide:
+
+====================================
+Modifier Definition Developers Guide
+====================================
+
+Modifier definitions files represent standardized objects which can be composed
```
Comment (rfbgo) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]:
```suggestion
Modifier definition files represent standardized objects which can be composed
```

File: lib/ramble/docs/dev_guides/modifier_dev_guide.rst [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+Before writing a modifier definition file, it is helpful to determine what
+aspects of the modification need to be grouped together.
+
+It can be useful to think of a modifier as some pattern you would like to apply
+to many experiments, that doesn't actually have anything to do with the
+behavior of an application directly. Depending on your goal for modifiers and
+the application of these patterns, you could write one modifier to apply all of
+the changes, or encapsulate a smaller set of changes into many modifiers.
+
+Some of the information you need to gather before writing your modifier includes:
```
Comment (rfbgo) [inline lib/ramble/docs/dev_guides/modifier_dev_guide.rst]:
```suggestion
Some of the information you need to gather before writing your modifier include:
```
Mix of plural and singular here currently :)

**Review by rfbgo — CHANGES_REQUESTED** [review]

**Review by rfbgo — APPROVED** [review]

**Review by linsword13 — APPROVED** [review]

**Pull request merged**